src/main/java/com/amazon/opendistroforelasticsearch/sql/esdomain/mapping/FieldMapping.java

```diff
@@ -0,0 +1,134 @@
+    private final Map<String, FieldMappingMetaData> typeMappings;
+
+    /** Map of field name to Field object that specified in query explicitly */
+    private final Map<String, Field> specifiedFieldByNames;
+
+
+    public FieldMapping(String fieldName) {
+        this(fieldName, emptyMap(), emptyMap());
+    }
+
+    public FieldMapping(String fieldName,
+                        Map<String, FieldMappingMetaData> typeMappings,
+                        Map<String, Field> specifiedFieldByNames) {
+
+        this.fieldName = fieldName;
+        this.typeMappings = typeMappings;
+        this.specifiedFieldByNames = specifiedFieldByNames;
+    }
+
+    /**
+     * Is field specified explicitly in query
+     * @return true if specified
+     */
+    public boolean isSpecified() {
+        return specifiedFieldByNames.containsKey(fieldName);
+    }
+
+    /**
+     * Is field a property field, which means either object field or nested field.
+     * @return true for property field
+     */
+    public boolean isPropertyField() {
+        int numOfDots = StringUtils.countMatches(fieldName, '.');
+        return numOfDots > 1 || (numOfDots == 1 && !isMultiField());
+    }
+
+    /**
+     * Path of property field, for example "employee" in "employee.manager"
+     * @return path of property field
+     */
+    public String path() {
+        int lastDot = fieldName.lastIndexOf(".");
+        return fieldName.substring(0, lastDot);
+    }
+
+    /**
+     * Verify if property field matches wildcard pattern specified in query
+     * @return true if matched
+     */
+    public boolean isWildcardSpecified() {
+        return specifiedFieldByNames.containsKey(path() + ".*");
+    }
+
+    /**
+     * Is field a/in multi-field, for example, field "a.keyword" in field "a"
+     * @return true for multi field
+     */
+    public boolean isMultiField() {
+        return fieldName.endsWith(".keyword");
+    }
+
+    /**
+     * Is field meta field, such as _id, _index, _source etc.
+     * @return true for meta field
+     */
+    public boolean isMetaField() {
+        return fieldName.startsWith("_");
+    }
+
+    /**
+     * Used to retrieve the type of fields from metaData map structures for both regular and nested fields
+     */
+    @SuppressWarnings("unchecked")
+    public String type() {
+        FieldMappingMetaData metaData = typeMappings.get(fieldName);
+        Map<String, Object> source = metaData.sourceAsMap();
+        String[] fieldPath = fieldName.split("\\.");
+
+        /*
+         * When field is not nested the metaData source is fieldName -> type
+         * When it is nested or contains "." in general (ex. fieldName.nestedName) the source is nestedName -> type
+         */
+        Map<String, Object> fieldMapping;
+        if (fieldPath.length < 2) {
+            fieldMapping = (Map<String, Object>) source.get(fieldName);
+        } else {
+            fieldMapping = (Map<String, Object>) source.get(fieldPath[1]);
```

Comment:
`source.get(fieldPath[fieldPath.length - 1])` to make sure this works even when we support multiple levels of nesting.

Reply:
I guess other code may break even if this works...

Reply:
sorry for nitpicking, but you could do
```
fieldMapping = (Map<String, Object>) source.get(fieldPath.length < 2 ? fieldName : fieldPath[1]);
```
to have no duplicated code.

Reply:
> sorry for nitpicking, but you could do
> 
> ```
> fieldMapping = (Map<String, Object>) source.get(fieldPath.length < 2 ? fieldName : fieldPath[1]);
> ```
> 
> to have no duplicated code.

Sure, will do.